### ISSUE TITLE
asyncBatch/asyncMapBatch

### ISSUE THREAD

**Issue by conartist6** [sithmel/iter-tools#22]
So I noticed this because the `asyncMapBatch` test is flaky. The test is flaky because it measures the time taken to execute and asserts on it. On any shared CI environment execution time is going to be higher than on a local machine.

This is relatively trivial to fix, the problem is that `asyncBatch` doesn't behave as expected in the first place.

The function is defined as
```
asyncBatch(n, asyncIterable) { ...
```

Unfortunately it doesn't parallelize anything, which it appears it was meant to. It just does ` for await (const item of asyncIter(iterable))` which fully serializes the async operations.

`mapAsyncBatch` does an `asyncMap` on an `asyncMap` (intended? not certain.) The async mapper function does indeed get batched correctly, and that's the only thing that is tested.

I'm going to take a stab at making it work the way I believe it should but it'll be another breaking change which is unfortunate. We should probably consider not releasing all breaking changes as major releases right away. If we gathered them up over a period of time in an @next branch then they might be less disruptive to however many users actually keep iter-tools up to date.

**Comment by conartist6** [comment]
Let me make sure I have the intent of the function right: I see its use as power infinite-scroll style paging experiences.

I say this because it's implementation differs significantly from a pure worker pool, whose only goal is to finish __all__ items in an efficient manner. It also, as an iterator, doesn't support random access and so isn't terribly useful for normal paging.

I would assume a common usage might look like:
```
const user = {
  friends: [4, 100, 101, 100021, 112358, ... ] // hundreds of items, perhaps
}

const batcher = asyncBatch(25, map(id => fetch(`users/${id}`).then(res => res.json()), user.friends))

const page1Results = await batcher.next(); // data for each users fetched in parallel
for (const friend of page1Results) {
  renderFriend(friend);
}

const page2Results = await batcher.next()
// and so on and so forth
```

Is this correct?

**Comment by conartist6** [comment]
Other thoughts:
Batch should batch into iterables not arrays.
Batch's type definition is wrong.

It is currently:
```
declare function batch<T>(number: number, iterable: IterableLike<T>): Iterable<T>;
```
but should be:
```
declare function batch<T>(number: number, iterable: IterableLike<T>): Iterable<Iterable<T>>;
```

**Comment by conartist6** [comment]
OK, wake up call. I reread the async iterators proposal and realized I've sort of been hazy on some critical aspects of how they work. They return `Promise<{value, done}>` because you don't know if the iterator is done synchronously. This makes a great many things here make more sense. `asyncBatch` __must__ use `await ... for ... of` because there's no other way to consume async iterators (being iterator functions defined on Symbol.asyncIterator).

So I guess `asyncBatch` makes sense to me as is again. I just think that `asyncMapBatch` should become  something like `batchPromises`, with semantics described above.

**Comment by conartist6** [comment]
So `promiseBatch` would just be:
```
map(Promise.all, batch(number, promiseIterable))
```

I think this is more the power of the existing API showing itself than it is a method that needs to be part of the library. Perhaps it could be included in a "cookbook" section of the documentation meant to show off how to easily use the library to accomplish some high level goals?

**Comment by sithmel** [comment]
asyncMapBatch can go. I agree it was a mistake adding it, it is too high level and specific to belong here.

About batch and asyncBatch: I asked myself if it was worth having them output an iterable of iterables (like groupby). Maybe it is actually worth it ...

**Comment by conartist6** [comment]
Closed by #21
